src/main/java/io/quarkus/search/app/quarkiverseio/QuarkiverseIO.java

```diff
@@ -0,0 +1,132 @@
+package io.quarkus.search.app.quarkiverseio;
+
+import java.io.Closeable;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+import io.quarkus.search.app.entity.Guide;
+import io.quarkus.search.app.hibernate.InputProvider;
+import io.quarkus.search.app.indexing.reporting.FailureCollector;
+import io.quarkus.search.app.util.CloseableDirectory;
+
+import io.quarkus.logging.Log;
+
+import org.hibernate.search.util.common.impl.Closer;
+
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+
+public class QuarkiverseIO implements Closeable {
+
+    public static final String QUARKIVERSE_ORIGIN = "quarkiverse-hub";
+
+    private final FailureCollector failureCollector;
+
+    private final List<Guide> quarkiverseGuides = new ArrayList<>();
+    private final boolean enabled;
+    private final Path pages;
+    private final URI baseUri;
+    private final CloseableDirectory tempDir;
+
+    public QuarkiverseIO(boolean enabled, Path pages, URI baseUri, FailureCollector failureCollector,
+            CloseableDirectory tempDir) {
+        this.failureCollector = failureCollector;
+        this.enabled = enabled;
+        this.pages = pages;
+        this.baseUri = baseUri;
+        this.tempDir = tempDir;
+    }
+
+    public void parseGuides() {
+        try (Stream<Path> stream = Files.list(pages)) {
+            List<Path> guideDirectories = stream.filter(Files::isDirectory)
+                    .filter(dir -> dir.getFileName().toString().startsWith("quarkus"))
+                    .toList();
+            for (Path directory : guideDirectories) {
+                Files.walkFileTree(
+                        directory, new SimpleFileVisitor<>() {
+                            @Override
+                            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                                if (file.getFileName().toString().endsWith(".html")) {
+                                    quarkiverseGuides.add(readGuide(file));
```

Comment:
Am I reading this right, we're putting all Quarkiverse guides (more than 100MB, from what I read elsewhere) in memory in one go?

Would you mind moving to a stream-based approach instead? :)
`Files.walk`/`Files.find` may help (they return a `Stream`).

Reply:
Entire quarkiverse unpacked is ~250 🙈 but there are images and stuff, we keep much less in memory 😄 but I agree this would be a better approach. Will do 😃